haystack/dataclasses/answer.py

```diff
@@ -1,25 +1,139 @@
-from typing import Any, Dict, List, Optional
-from dataclasses import dataclass
+import io
+from typing import Any, Dict, List, Optional, Protocol, runtime_checkable
+from dataclasses import dataclass, field, asdict
+
+from pandas import DataFrame, read_json
+
+from haystack.core.serialization import default_from_dict, default_to_dict
 from haystack.dataclasses.document import Document
 
 
-@dataclass(frozen=True)
-class Answer:
+@runtime_checkable
+@dataclass
+class Answer(Protocol):
     data: Any
     query: str
-    metadata: Dict[str, Any]
+    meta: Dict[str, Any]
+
+    def to_dict(self) -> Dict[str, Any]:
+        ...
+
+    @classmethod
+    def from_dict(cls, data: Dict[str, Any]) -> "Answer":
+        ...
+
+
+@dataclass
+class ExtractedAnswer:
+    query: str
+    score: float
+    data: Optional[str] = None
+    document: Optional[Document] = None
+    context: Optional[str] = None
+    document_offset: Optional["Span"] = None
+    context_offset: Optional["Span"] = None
+    meta: Dict[str, Any] = field(default_factory=dict)
+
+    @dataclass
+    class Span:
+        start: int
+        end: int
+
+    def to_dict(self) -> Dict[str, Any]:
+        document = self.document.to_dict(flatten=False) if self.document is not None else None
+        document_offset = asdict(self.document_offset) if self.document_offset is not None else None
+        context_offset = asdict(self.context_offset) if self.context_offset is not None else None
+        return default_to_dict(
+            self,
+            data=self.data,
+            query=self.query,
+            document=document,
+            context=self.context,
+            score=self.score,
+            document_offset=document_offset,
+            context_offset=context_offset,
+            meta=self.meta,
+        )
 
+    @classmethod
+    def from_dict(cls, data: Dict[str, Any]) -> "ExtractedAnswer":
+        init_params = data.get("init_parameters", {})
+        if (doc := init_params.get("document")) is not None:
+            data["init_parameters"]["document"] = Document.from_dict(doc)
 
-@dataclass(frozen=True)
-class ExtractedAnswer(Answer):
-    data: Optional[str]
-    document: Optional[Document]
-    probability: float
```

Comment:
I ran the e2e tests now so that you can see where any other e2e test might fail: https://github.com/deepset-ai/haystack/actions/runs/7170750987/job/19524201798

Reply:
Tutorials and examples seem to be unaffected.